### PR TITLE
Updating xunit version

### PR DESCRIPTION
**Pull request by moozzyk**
Allows running tests inside VS and from build.cmd
~Functional tests will come separately~

### CODE REVIEW

File: tests/Microsoft.AspNet.SignalR.Tests/packages.config [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Moq" version="4.0.10827" />
+  <package id="Moq" version="4.0.10827" requireReinstallation="true" />
   <package id="Newtonsoft.Json" version="6.0.4" targetFramework="net45" />
   <package id="xunit" version="1.9.1" targetFramework="net40" />
   <package id="xunit.extensions" version="1.9.1" targetFramework="net45" />
```
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Tests/packages.config]: these should be 2.2.0

File: tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/UnitTestApp.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
+﻿using System;
```
Comment (mikaelm12) [inline tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/UnitTestApp.xaml.cs]: nit: Header 😄

File: src/Microsoft.AspNet.SignalR.Redis/RedisScaleoutConfiguration.cs [hunk 3]
```diff
@@ -31,8 +31,11 @@ public RedisScaleoutConfiguration(string connectionString, string eventKey)
             }
 
             ConnectionString = connectionString;
-            var options = ConfigurationOptions.Parse(connectionString);
-            Database = options.DefaultDatabase.GetValueOrDefault(0);
+            if (connectionString.Length > 0)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNet.SignalR.Redis/RedisScaleoutConfiguration.cs]: `!string.IsNullOrEmpty`

**Comment by BrennanConroy** [comment]
:shipit:

File: src/Microsoft.AspNet.SignalR.Core/Messaging/MessageBus.cs [hunk 4]
```diff
@@ -152,7 +152,7 @@ public virtual Task Publish(Message message)
             if (Topics.TryGetValue(message.Key, out topic))
             {
                 topic.Store.Add(message);
-                ScheduleTopic(message.Key, topic);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.Core/Messaging/MessageBus.cs]: Unrelated change, but seems fine.

File: tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj [hunk 5]
```diff
@@ -22,6 +22,8 @@
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
     <MSBuildCommunityTasksPath Condition="$(MSBuildCommunityTasksPath) == ''">..\Build</MSBuildCommunityTasksPath>
+    <TargetFrameworkProfile />
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj]: Remove?

File: tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj [hunk 6]
```diff
@@ -62,6 +64,7 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\packages\Owin.1.0\lib\net40\Owin.dll</HintPath>
     </Reference>
+    <Reference Include="System.Data.DataSetExtensions" />
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj]: Remove

File: tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Web.config [hunk 7]
```diff
@@ -1,20 +1,28 @@
...
+    The following attributes can be set on the <httpRuntime> tag.
+      <system.Web>
+        <httpRuntime targetFramework="4.5.2" />
+      </system.Web>
+  -->
   <system.web>
-    <compilation debug="true" targetFramework="4.5" />
-    <httpRuntime targetFramework="4.5" />
+    <compilation debug="true" targetFramework="4.5.2"/>
+    <httpRuntime targetFramework="4.5"/>
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Web.config]: Remove this? Do we want to target net45?

File: tests/Microsoft.AspNet.SignalR.Client.Portable.Tests/packages.config [hunk 8]
```diff
@@ -1,8 +1,41 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Microsoft.NETCore.Platforms" version="1.0.1" targetFramework="net45" />
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.Portable.Tests/packages.config]: Maybe we should switch to `PackageReference` eventually.

File: tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/Package.appxmanifest [hunk 9]
```diff
@@ -0,0 +1,45 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Package
+  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
+  xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest"
+  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
+  IgnorableNamespaces="uap mp">
+
+  <Identity Name="1e8e9073-5284-4ce7-94c0-7d32ca415e4f"
+            Publisher="CN=pawelka"
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/Package.appxmanifest]: This doesn't look right.

File: tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs [hunk 10]
```diff
@@ -65,9 +65,10 @@ public void OpenCalledOnConnectionRestored()
                 }
             });
 
-            // Creating an instance to invoke the constructor which starts the connection
             var instance = redisMessageBus.Object;
+            await instance.ConnectWithRetry();
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs]: Why this change?

File: tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs [hunk 11]
```diff
@@ -105,8 +106,11 @@ public void RestoreLatestValueForKeyCalledOnConnectionRestored()
             });
 
             var redisMessageBus = new RedisMessageBus(GetDependencyResolver(), new RedisScaleoutConfiguration(String.Empty, String.Empty),
-            redisConnection.Object, true);
+            redisConnection.Object, connectAutomatically: false);
 
+            await redisMessageBus.ConnectWithRetry();
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs]: Same question as above

File: tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs [hunk 12]
```diff
@@ -113,7 +113,7 @@ public void ContinuePreservedCulturePreservesCulture()
                 (task, continuation) => task.ContinueWithPreservedCulture(_ => continuation()));
         }
 
-        [Fact]
+        [Fact(Skip = "Started failing after upgrading xunit from 1.9.1 to 2.2.0")]
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs]: Can you file a bug to follow up on this.

File: tests/Microsoft.AspNet.SignalR.Tests/packages.config [hunk 13]
```diff
@@ -1,7 +1,12 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Moq" version="4.0.10827" />
+  <package id="Moq" version="4.0.10827" requireReinstallation="true" />
```
Comment (davidfowl) [inline tests/Microsoft.AspNet.SignalR.Tests/packages.config]: requireReinstallation? Can you reinstall to make sure nuget gets the right assemblies in the project file?

File: src/Microsoft.AspNet.SignalR.Core/Messaging/MessageBus.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.AspNet.SignalR.Core/Messaging/MessageBus.cs]: ScheduleTopic had unused parameter. Needed to fix this because code would not compile due to an error. Not sure why it did not show up before.

File: src/Microsoft.AspNet.SignalR.Redis/RedisScaleoutConfiguration.cs (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.AspNet.SignalR.Redis/RedisScaleoutConfiguration.cs]: there is already a null check above.

File: tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj (hunk 5 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj]: Don't know - VS added this. I can try.

File: tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj (hunk 6 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Client.JS.Tests/Microsoft.AspNet.SignalR.Client.JS.Tests.csproj]: OK

File: tests/Microsoft.AspNet.SignalR.Client.Portable.Tests/packages.config (hunk 8 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Client.Portable.Tests/packages.config]: Not sure. At the moment you can't open this project with VS2017 due to store and WinRT projects and I don't know if VS2015 supports PackageReference.

File: tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/Package.appxmanifest (hunk 9 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Client.UWP.Tests/Package.appxmanifest]: This is a test certificate. I think it can be changed.

File: tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs (hunk 10 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs]: Because we now don't raise the ConnectionRestored event when the connection is in the open state (to avoid duplicate events) this test stopped working. It relied on the fact that ConnectionRestored/ConnectionFailed will be fired regardless of the state the connection is in.

File: tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs (hunk 11 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Redis.Tests/RedisMessageBusFacts.cs]: Same answer as above

File: tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs (hunk 12 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs]: OK. We actually saw exactly the same in SignalR-Server https://github.com/aspnet/SignalR-Server/blob/9680e9c2c185a19604390003a6f10489d0538a56/test/Microsoft.AspNetCore.SignalR.Server.Tests/TaskAsyncHelperFacts.cs#L127-L128. I once briefly looked at it but couldn't find anything obvious. Maybe I should ask xUnit folks if they have any ideas.

File: tests/Microsoft.AspNet.SignalR.Tests/packages.config (hunk 13 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Tests/packages.config]: Not sure why VS added it - maybe due to retargeting the project to net452 - not the `targetframework` is missing. I will try to update and unify versions of Moq.

File: tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs (hunk 12 above)
Comment (moozzyk) [inline tests/Microsoft.AspNet.SignalR.Tests/TaskAsyncHelperFacts.cs]: https://github.com/SignalR/SignalR/issues/3901

File: tests/Microsoft.AspNet.SignalR.Redis.Tests/Microsoft.AspNet.SignalR.Redis.Tests.csproj [hunk 14]
```diff
@@ -10,11 +9,13 @@
...
     <AssemblyName>Microsoft.AspNet.SignalR.Redis.Tests</AssemblyName>
-    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
-    <NuGetPackageImportStamp>25f78d1d</NuGetPackageImportStamp>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
+    <TargetFrameworkProfile />
```
Comment (anurse) [inline tests/Microsoft.AspNet.SignalR.Redis.Tests/Microsoft.AspNet.SignalR.Redis.Tests.csproj]: Tidy up these empty properties

File: tests/Microsoft.AspNet.SignalR.SqlServer.Tests/Microsoft.AspNet.SignalR.SqlServer.Tests.csproj [hunk 15]
```diff
@@ -8,10 +8,11 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>Microsoft.AspNet.SignalR.SqlServer.Tests</RootNamespace>
     <AssemblyName>Microsoft.AspNet.SignalR.SqlServer.Tests</AssemblyName>
-    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
+    <TargetFrameworkProfile />
```
Comment (anurse) [inline tests/Microsoft.AspNet.SignalR.SqlServer.Tests/Microsoft.AspNet.SignalR.SqlServer.Tests.csproj]: Clean up

File: tests/Microsoft.AspNet.SignalR.Tests/Microsoft.AspNet.SignalR.Tests.csproj [hunk 16]
```diff
@@ -13,8 +14,10 @@
     <FileAlignment>512</FileAlignment>
     <RestorePackages>true</RestorePackages>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
-    <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
     <TargetFrameworkProfile />
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (anurse) [inline tests/Microsoft.AspNet.SignalR.Tests/Microsoft.AspNet.SignalR.Tests.csproj]: Clean up

File: tests/Microsoft.AspNet.SignalR.Tests/packages.config (hunk 13 above)
Comment (anurse) [inline tests/Microsoft.AspNet.SignalR.Tests/packages.config]: Yeah, this is intended to be a flag for the UI to prompt you to reinstall the package, at least that's what I thought it was.

**Comment by moozzyk** [comment]
:up::date:

**Comment by moozzyk** [comment]
Converted functional tests - it is not as big as I thought it would be

**Comment by moozzyk** [comment]
:watch:
